### PR TITLE
KAFKA-16381 use volatile to guarantee KafkaMetric#config visibility across threads

### PR DESCRIPTION
**Pull request by johnnychhsu**
## Context
In KafkaMetirc.java, the config getter is 

```
@Override
public MetricName metricName()
{ return this.metricName; }
```

and there is a setter 

```
public void config(MetricConfig config) {
    synchronized (lock) {    
         this.config = config;  
     }
}
```

Since it's possible to set and get in the mean time, we should consider this case for getting.

Refer to [KAFKA-16381](https://issues.apache.org/jira/browse/KAFKA-16381).

## Solution
Set the config as volatile, which can avoid the case, and it's also lighter than adding a lock in getter.

## Test
```
./gradlew cleanTest core:test --tests ConnectionQuotasTest --tests ControllerMutationQuotaTest stream:test --tests StandbyTaskTest --tests StreamTaskTest
```
and it passed

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by vamossagar12 — COMMENTED** [review]
Thanks for the PR @johnnychhsu . IMHO, we don't need to acquire a lock in this case as this is just the read part. The config object is set once from the constructor which is what is being returned here. Now, if some other thread is trying to update the config by using `config(MetricConfig config)`, then unless that operation doesn't go through, we can keep returning the unchanged config object. WDYT?

**Comment by johnnychhsu** [comment]
Thanks for the review @vamossagar12 .
let me rephrase your comment to make sure that I understand you correctly.
> Now, if some other thread is trying to update the config by using config(MetricConfig config), then unless that operation 
> doesn't go through, we can keep returning the unchanged config object.
So you mean only if the `config(MetricConfig config)` works, then we need to return the updated config, right?

I feel it's better to have it here, because if someone update the config, then immediately use the config as a criteria to do something else, but the config has not updated yet or somehow fails, then this could lead to a potential problem. 
wdyt?

**Comment by vamossagar12** [comment]
> So you mean only if the config(MetricConfig config) works, then we need to return the updated config, right?

yes, that's correct. Any other thread should see the updated state of the shared variables in the synchronised block only after the lock is released. And come to think of it, that is already being achieved without the synchronised block in the `config()` method.

The changes you have added, provide the same guarantees as whatever exist today but at the expense of adding a lock. IF we want to make the value of the `config` object visible immediately to other threads, we could consider making it `volatile` but i am not sure if we really need it. WDYT?

**Comment by ijuma** [comment]
I agree that the current behavior doesn't look right, but adding a lock is also too heavyweight. `volatile` seems reasonable.

**Comment by johnnychhsu** [comment]
thanks @vamossagar12 @ijuma for the comments.
That's a fair point, I agree that a lock is heavy, and `volatile` should be good enough for this. Let me change this, thanks!

**Review by vamossagar12 — APPROVED** [review]
LGTM. Thanks for the change.

**Review by chia7712 — APPROVED** [review]

**Comment by vamossagar12** [comment]
@johnnychhsu , you could probably update the PR description to reflect the latest changes.

**Comment by johnnychhsu** [comment]
@vamossagar12 thanks for the reminder.
Updated the PR description to reflect the latest changes.

**Comment by chia7712** [comment]
```
./gradlew cleanTest :trogdor:test --tests CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated :connect:mirror:test --tests MirrorConnectorsWithCustomForwardingAdminIntegrationTest.testCreatePartitionsUseProvidedForwardingAdmin :core:test --tests DelegationTokenEndToEndAuthorizationWithOwnerTest.testProduceConsumeTopicAutoCreateTopicCreateAcl --tests DelegationTokenEndToEndAuthorizationWithOwnerTest.testNoDescribeProduceOrConsumeWithoutTopicDescribeAcl --tests SocketServerTest.testControlPlaneTakePrecedenceOverInterBrokerListenerAsPrivilegedListener --tests SocketServerTest.testZeroMaxConnectionsPerIp --tests SocketServerTest.testStagedListenerShutdownWhenConnectionQueueIsFull --tests SocketServerTest.testStagedListenerStartup --tests SocketServerTest.testControlPlaneAsPrivilegedListener --tests SocketServerTest.testInterBrokerListenerAsPrivilegedListener --tests DynamicBrokerReconfigurationTest.testThreadPoolResize --tests ReplicaManagerTest.testRemoteFetchExpiresPerSecMetric --tests SaslPlainPlaintextConsumerTest.testClusterResourceListener --tests TransactionsTest.testFencingOnSend --tests TransactionsTest.testCommitTransactionTimeout --tests LogDirFailureTest.testIOExceptionDuringLogRoll --tests LogDirFailureTest.testIOExceptionDuringCheckpoint :clients:test --tests SelectorTest.testConnectionsByClientMetric --tests Tls12SelectorTest.testConnectionsByClientMetric --tests SelectorTest.testInboundConnectionsCountInConnectionCreationMetric --tests Tls12SelectorTest.testInboundConnectionsCountInConnectionCreationMetric --tests SelectorTest.testMuteOnOOM --tests Tls12SelectorTest.testMuteOnOOM
```
the failed tests are unrelated and they pass on my local
